app/services/fact_importer.rb

```diff
@@ -61,8 +61,14 @@ def fact_name_class
 
   def delete_removed_facts
     ActiveSupport::Notifications.instrument "fact_importer_deleted.foreman", :host_id => host.id, :host_name => host.name, :facts => facts, :deleted => [] do |payload|
-      # deletes all facts using a single SQL query (with inner query)
-      payload[:count] = @counters[:deleted] = FactValue.joins(:fact_name).where(:host => host, 'fact_names.type' => fact_name_class).where.not('fact_names.name' => facts.keys).delete_all
+      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'
+        # MySQL does not handle delete with inner query correctly (slow) so we will do N+1 here on purpose
+        ids = FactValue.joins(:fact_name).where(:host => host, 'fact_names.type' => fact_name_class).where.not('fact_names.name' => facts.keys).pluck(:id)
```

Comment:
Instead of repeating the same query building with a different last method, how about saving the ` FactValue.joins(:fact_name).where(:host => host, 'fact_names.type' => fact_name_class).where.not('fact_names.name' => facts.keys)` part in a variable and just calling a different method on it?